File: leetcode/100-999/leetcode133.cc
#include <iostream>
#include <queue>
#include <set>
#include <vector>
using namespace std;
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node()
    {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val)
    {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors)
    {
        val = _val;
        neighbors = _neighbors;
    }
};
class Solution {
public:
    int used[105];
    Node* rec[105];
    Node* dfs(Node* node)
    {
        if (node == nullptr)
            return nullptr;
        Node* neo = new Node(node->val);
        used[node->val] = 1;
        rec[node->val] = neo;
        for (Node* nei : node->neighbors) {
            if (used[nei->val] == 1) {
                neo->neighbors.push_back(rec[nei->val]);
                continue;
            }
            neo->neighbors.push_back(dfs(nei));
        }
        return neo;
    }
    Node* cloneGraph(Node* node)
    {
        if (node == nullptr)
            return nullptr;
        return dfs(node);
    }
};
int main(void) { return 0; }

File: leetcode/100-999/leetcode133.py
# Definition for a Node.
class Node:
    def __init__(self, val=0, neighbors=None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []


class Solution:
    def cloneGraph(self, node: "Node") -> "Node":
        if not node:
            return None
        queue = [node]
        dic = {node: Node(val=node.val)}
        while queue:
            cur = queue.pop(0)
            for neighbor in cur.neighbors:
                if not dic.get(neighbor):
                    dic[neighbor] = Node(val=neighbor.val)
                    queue.append(neighbor)
        vist = set()
        queue.append(node)
        while queue:
            cur = queue.pop(0)
            vist.add(cur)
            dic[cur].neighbors = [dic[neighbor] for neighbor in cur.neighbors]
            for neighbor in cur.neighbors:
                if neighbor not in vist:
                    queue.append(neighbor)
        return dic[node]
